### PR TITLE
Add Cognito domain creation and expose OAuth to frontend config

### PR DESCRIPTION
**Pull request by awsluja**
Issue #, if available:

Description of changes:
Original PR: https://github.com/aws-amplify/amplify-backend/pull/756
Reverted: https://github.com/aws-amplify/amplify-backend/pull/838
This PR is to fix the reverted PR, after some e2e failures were found.

This change enables the creation of a Cognito domain given the specified 'domainPrefix'.
It allows users to create, update, or delete the domain after initial deployments (with a slight workaround).

Also, exposes the OAuth settings to the frontend config.

social providers
oauth (domain, scope, signin/out urls, response type)

UX:
image

## Changes

**Corresponding docs PR, if applicable:**

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts [hunk 1]
```diff
@@ -19,6 +19,9 @@ export const auth = defineAuth({
       //   clientId: secret('LOGINWITHAMAZON_CLIENT_ID'),
       //   clientSecret: secret('LOGINWITHAMAZON_CLIENT_SECRET'),
       // }
+      // configure callback and logout URLs
+      callbackUrls: ['https://redirect.com'],
+      logoutUrls: ['https://logout.com'],
```
Comment (edwardfoyle) [inline packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts:24]: Is this required in all cases now? Ideally our getting started template should create the "simplest" auth config possible. Maybe we need to make the entire `externalProviders` block a comment

**Review by sobolk — APPROVED** [review]

File: packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts (hunk 1 above)
Comment (awsluja) [reply packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts:24]: yeah we have a requirement that users provide these if they are specifying external providers

Comment (edwardfoyle) [reply packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts:24]: Right, but should the basic getting started template include `externalProviders` at all?

Comment (sobolk) [reply packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts:24]: another interesting angle here:
how do I define different url per branch/sandbox? (does this matter?)

Comment (edwardfoyle) [reply packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts:24]: I think what we decided for that would be something like `logoutUrls: process.env.AWS_BRANCH === 'prod' ? ['https://produrl.com'] : ['http://localhost:3000']` (the AWS_BRANCH env var is automatically set by amplify hosting builds)

Comment (awsluja) [reply packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts]: I've updated to comment out external providers, and use http://localhost:3000 as the default for both, see https://github.com/aws-amplify/amplify-backend/pull/856/commits/20975bcb87eadc0fe1eb59ae90b636cd03a84a4f

File: packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts [hunk 2]
```diff
@@ -9,20 +9,20 @@ export const auth = defineAuth({
...
-      // loginWithAmazon: {
-      //   clientId: secret('LOGINWITHAMAZON_CLIENT_ID'),
-      //   clientSecret: secret('LOGINWITHAMAZON_CLIENT_SECRET'),
-      // }
-      // configure callback and logout URLs
-      callbackUrls: ['https://redirect.com'],
-      logoutUrls: ['https://logout.com'],
-    },
+    // externalProviders: {
+    /**
```
Comment (edwardfoyle) [inline packages/create-amplify/templates/basic-auth-data/amplify/auth/resource.ts:13]: super duper nit: make this either one block comment or all single line comments

**Review by edwardfoyle — APPROVED** [review]

**Review by sobolk — APPROVED** [review]

**Pull request merged**